### PR TITLE
Adding EventSource version of original WCF/WF ETW provider

### PR DESCRIPTION
**Pull request by dmetzgar**
This is a first stab at issue #38

It includes the WCF/WF ETW provider turned into an EventSource and wires up some of the events.

### CODE REVIEW

File: src/System.Private.ServiceModel/src/System.Private.ServiceModel.csproj [hunk 1]
```diff
@@ -17,16 +17,17 @@
...
-     <DefineConstants>$(DefineConstants);FEATURE_CORECLR</DefineConstants>
-    </PropertyGroup>
+    <DefineConstants>$(DefineConstants);FEATURE_CORECLR</DefineConstants>
+  </PropertyGroup>
   <ItemGroup>
     <Compile Include="$(CommonPath)\System\NotImplemented.cs">
       <Link>Common\System\NotImplemented.cs</Link>
     </Compile>
     <Compile Include="$(MsBuildThisFileDirectory)\**\*.cs" />
+    <Compile Include="Internals\WcfEventSource.cs" />
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System.Private.ServiceModel.csproj]: Remove this line as it causes the file to be seen twice. VS isn't very smart about avoiding adding new files to the csproj when it's in a path that's wildcard included.

File: src/System.Private.ServiceModel/src/project.lock.json [hunk 2]
```diff
@@ -3490,6 +3490,7 @@
       "System.Diagnostics.Contracts >= 4.0.0",
       "System.Diagnostics.Debug >= 4.0.10",
       "System.Diagnostics.Tools >= 4.0.0",
+      "System.Diagnostics.Tracing >= 4.0.20",
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/project.lock.json]: You need to revert this file change and run `msbuild /t:RestorePackages /p:LockDependencies=true`.

**Comment by roncain** [comment]
I buddy-built this and this innerloop test fails:
08:28:20.462    22>RunTestsForProject: (TargetId:122)
                        BasicHttpBindingTest.ReaderQuotas_Property_Set_Null_Throws [FAIL] (TaskId:105)
                           Assert.Throws() Failure (TaskId:105)
                           Expected: typeof(System.ArgumentNullException) (TaskId:105)
                           Actual:   typeof(System.NullReferenceException): Object reference not set to an instance of an object. (TaskId:105)
                           Stack Trace: (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\Internals\System\Runtime\Diagnostics\EtwDiagnosticTrace.cs(366,0): at System.Runtime.Diagnostics.EtwDiagnosticTrace.WriteXmlElementString(XmlWriter xml, String localName, String value, Int32& remainingLength) (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\Internals\System\Runtime\Diagnostics\EtwDiagnosticTrace.cs(248,0): at System.Runtime.Diagnostics.EtwDiagnosticTrace.WriteExceptionToTraceString(XmlWriter xml, Exception exception, Int32 remainingLength, Int32 remainingAllowedRecursionDepth) (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\Internals\System\Runtime\Diagnostics\EtwDiagnosticTrace.cs(210,0): at System.Runtime.Diagnostics.EtwDiagnosticTrace.ExceptionToTraceString(Exception exception, Int32 maxTraceStringLength) (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\Internals\TraceCore.Designer.cs(136,0): at System.Runtime.TraceCore.ThrowingException(EtwDiagnosticTrace trace, String param0, String param1, Exception exception) (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\Internals\System\Runtime\ExceptionTrace.cs(225,0): at System.Runtime.ExceptionTrace.TraceException[TException](TException exception, String eventSource) (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\Internals\System\Runtime\ExceptionTrace.cs(214,0): at System.Runtime.ExceptionTrace.TraceException[TException](TException exception) (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\Internals\System\Runtime\ExceptionTrace.cs(176,0): at System.Runtime.ExceptionTrace.ArgumentNull(String paramName) (TaskId:105)
                              C:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\HttpBindingBase.cs(115,0): at System.ServiceModel.HttpBindingBase.set_ReaderQuotas(XmlDictionaryReaderQuotas value) (TaskId:105)
                              C:\OSS\wcf\src\System.ServiceModel.Http\tests\ServiceModel\BasicHttpBindingTest.cs(270,0): at BasicHttpBindingTest.<>c__DisplayClass20_0.<ReaderQuotas_Property_Set_Null_Throws>b__0() (TaskId:105)
                     Finished:    System.ServiceModel.Http.Tests (TaskId:105)

**Comment by roncain** [comment]
In WcfEventSource, it would be great if all the hard-coded integer event ID's could be pulled out into private constants or an enum.  As hard-coded values it's hard to know what else might depend on them being the values they are.

**Comment by dmetzgar** [comment]
@roncain That's an interesting bug with the tests. What's weird is that it comes from EtwDiagnosticTrace.WriteExceptionToTraceString where it creates a list of Tuples based on the exception and then cycles through them to write them to XML. Since it broke in the WriteXmlElementString I'm assuming one of the tuple items is null but the exception is not null. The suspicious call for me is to exception.StackTrace. I'm wondering if .StackTrace could be returning null.

**Comment by dmetzgar** [comment]
Updated the PR in response to the comments. I've tested the EventSource in WF for parity with framework but haven't added a test yet for WCF. Creating a unit test for the events is a bit tricky since any other WCF activity (like other unit tests running in parallel) could interfere with what's recorded making the test flaky.

**Comment by roncain** [comment]
LGTM but I would like one last look by @mconnew @iamjasonp or @SajayAntony .
The reason is that this PR also updates the project.lock.json files to newer versions, especially in the networking packages.  I've buddy tested this most recent PR iteration on 2 machines.  One machine passes all OuterLoop tests fine.  The other machine consistently fails one OuterLoop test.  It could be a flakiness introduced by the networking changes, so I am inclined to accept this PR and just watch that test's stability.
```
                        ExpectedExceptionTests.NotExistentHost_Throws_EndpointNotFoundException [FAIL] (TaskId:1802)
                           Assert.Throws() Failure (TaskId:1802)
                           Expected: typeof(System.ServiceModel.EndpointNotFoundException) (TaskId:1802)
                           Actual:   typeof(System.TimeoutException): The request channel timed out attempting to send after 00:00:10. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout. (TaskId:1802)
                           Stack Trace: (TaskId:1802)
                              e:\oss\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Channels\RequestChannel.cs(283,0): at System.ServiceModel.Channels.RequestChannel.<RequestAsync>d__34.MoveNext() (TaskId:1802)
                              --- End of stack trace from previous location where exception was thrown --- (TaskId:1802)
                                 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) (TaskId:1802)
                                 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) (TaskId:1802)
                                 at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult() (TaskId:1802)
 ```

**Comment by iamjasonp** [comment]
LGTM 

Clean built on my (not-used-for-dev-work) machine, and all inner and outerloop tests pass for me.

**Pull request merged**